src/sentry/relay/config/metric_extraction.py

```diff
@@ -409,42 +409,56 @@ def convert_widget_query_to_metric(
         return metrics_specs
 
-    for aggregate in widget_query.aggregates:
-        metrics.incr(
-            "on_demand_metrics.before_widget_spec_generation",
-            tags={"prefilling": prefilling},
+    aggregates = widget_query.aggregates or []
+    groupbys = widget_query.columns
+
+    for aggregate in aggregates:
+        metrics_specs += _generate_metric_specs(
+            aggregate, widget_query, project, prefilling, groupbys
         )
-        if results := _convert_aggregate_and_query_to_metrics(
-            project,
-            # there is an internal check to make sure we extract metrics only for performance dataset
-            # however widgets do not have a dataset field, so we need to pass it explicitly
-            Dataset.PerformanceMetrics.value,
-            aggregate,
-            widget_query.conditions,
-            None,
-            prefilling,
-            groupbys=widget_query.columns,
-            spec_type=MetricSpecType.DYNAMIC_QUERY,
-        ):
-            for spec in results:
-                _log_on_demand_metric_spec(
-                    project_id=project.id,
-                    spec_for="widget",
-                    spec=spec,
-                    id=widget_query.id,
-                    field=aggregate,
-                    query=widget_query.conditions,
-                    prefilling=prefilling,
-                )
-                metrics.incr(
-                    "on_demand_metrics.on_demand_spec.for_widget",
-                    tags={"prefilling": prefilling},
-                )
-                metrics_specs.append(spec)
 
     return metrics_specs
 
 
+def _generate_metric_specs(
+    aggregate: str,
+    widget_query: DashboardWidgetQuery,
+    project: Project,
+    prefilling: bool,
+    groupbys: Sequence[str] | None = None,
+) -> list[HashedMetricSpec]:
+    metrics_specs = []
+    metrics.incr("on_demand_metrics.before_widget_spec_generation")
+    if results := _convert_aggregate_and_query_to_metrics(
+        project,
+        # there is an internal check to make sure we extract metrics only for performance dataset
+        # however widgets do not have a dataset field, so we need to pass it explicitly
+        Dataset.PerformanceMetrics.value,
+        aggregate,
+        widget_query.conditions,
+        None,
+        prefilling,
+        groupbys=groupbys,
+        spec_type=MetricSpecType.DYNAMIC_QUERY,
+    ):
+        for spec in results:
+            _log_on_demand_metric_spec(
+                project_id=project.id,
+                spec_for="widget",
+                spec=spec,
+                id=widget_query.id,
+                field=aggregate,
+                query=widget_query.conditions,
+                prefilling=prefilling,
+            )
+            metrics.incr(
+                "on_demand_metrics.on_demand_spec.for_widget",
+                tags={"prefilling": prefilling},
+            )
+            metrics_specs.append(spec)
```

Comment:
This block is untouched:

<img width="395" alt="image" src="https://github.com/getsentry/sentry/assets/44410/abca05eb-53ab-44cd-8870-8e047642364c">